### PR TITLE
[ticket/15294] Fix session_gc() selecting expired sessions for unique users

### PR DESCRIPTION
**Pull request by rxu**
Checklist:

- [x] Correct branch: master for new features; 3.3.x & 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html), [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

<a href="https://tracker.phpbb.com/browse/PHPBB3-15294">PHPBB3-15294</a>.

### CODE REVIEW

**Comment by rxu** [comment]
I'm not sure if we need to select and update info (last session page, user last visit) in the users table for anonymous user here.

File: phpBB/phpbb/session.php [hunk 1]
```diff
@@ -954,72 +954,71 @@ function session_gc()
...
-
-		while ($row = $db->sql_fetchrow($result))
-		{
-			$sql = 'UPDATE ' . USERS_TABLE . '
-				SET user_lastvisit = ' . (int) $row['recent_time'] . ", user_lastpage = '" . $db->sql_escape($row['session_page']) . "'
-				WHERE user_id = " . (int) $row['session_user_id'];
-			$db->sql_query($sql);
+		// Delete all expired sessions
+		$sql = 'DELETE FROM ' . SESSIONS_TABLE . '
+			WHERE session_time < ' . ($this->time_now - $config['session_length']);
```
Comment (mrgoldy) [inline phpBB/phpbb/session.php]: Above you cast this variable to `(int)`, might want to do it here aswell.

File: phpBB/phpbb/session.php [hunk 2]
```diff
@@ -954,72 +954,71 @@ function session_gc()
...
+			case 'sqlite3':
+			case 'mysqli':
+				$sql = 'UPDATE ' . USERS_TABLE . " AS u,
+					$sql_select
+					SET u.user_lastvisit = s3.recent_time, u.user_lastpage = s3.session_page
+					WHERE u.user_id = s3.session_user_id";
+			break;
+
+			default:
+			$sql = 'UPDATE ' . USERS_TABLE . "
```
Comment (mrgoldy) [inline phpBB/phpbb/session.php]: Needs additional indentation

File: phpBB/phpbb/session.php [hunk 3]
```diff
@@ -954,72 +954,71 @@ function session_gc()
...
 			$this->time_now = time();
 		}
 
-		// Firstly, delete guest sessions
-		$sql = 'DELETE FROM ' . SESSIONS_TABLE . '
-			WHERE session_user_id = ' . ANONYMOUS . '
-				AND session_time < ' . (int) ($this->time_now - $config['session_length']);
+		// Get expired sessions, only most recent for each registered user
+		// Inner SELECT gets most recent expired sessions data for unique session_user_id
+		// Outer SELECT gets also session_page for them
```
Comment (mrgoldy) [inline phpBB/phpbb/session.php]: Perhaps it's better to make this a block comment, `/** ... */`.

**Comment by mrgoldy** [comment]
I believe the travis error is cause of the derived table alias, which could/should not have an `AS` infront.

**Comment by rxu** [comment]
> I believe the travis error is cause of the derived table alias, which could/should not have an `AS` infront.

I've tested locally on SQLite 3.20 and it's erroring both ways (with and withous `AS`). It seems that SQLite doesn't support `UPDATE` from `SELECT`, at least in such way, regrettably.

**Comment by 3D-I** [comment]
> Are there any SQLite experts around?

Not really, just found a related topic here: https://stackoverflow.com/questions/10027468/how-can-i-update-table-by-using-select-statement-in-sqlite#10027800

**Comment by rxu** [comment]
That would be possible in SQLite using CTE (common table expressions) but it's supported since SQLite 3.8.3 while phpBB 3.3 requires 3.6.15.

**Comment by 3D-I** [comment]
> 
> 
> That would be possible in SQLite using CTE (common table expressions) but it's supported since SQLite 3.8.3 while phpBB 3.3 requires 3.6.15.

Other way around could be rewrite that SQL to be DBAL compliant at once,
I mean diving the query and what not?

**Comment by 3D-I** [comment]
> I guess I'll better reset to the previous splitted solution.
That's what I meant.

**Comment by rxu** [comment]
Well, current approach looks better from the point it gets rid of executing `UPDATE` queries in a `while` loop. But there's no other choice for the moment I guess.

**Comment by 3D-I** [comment]
> 
> 
> Well, current approach looks better from the point it gets rid of executing `UPDATE` queries in a `while` loop. But there's no other choice for the moment I guess.

Or `foreach` maybe... wondering those approachs though will need a `sql->transaction` sort of loop.

**Comment by rxu** [comment]
> executing `UPDATE` queries in a `while` loop.

So, this is probably why the `$batch_size` was needed.

**Comment by rxu** [comment]
Still thinking on better implementation.
Please don't merge.

**Comment by rxu** [comment]
Tested on a live production board (v.3.2.8) for a week, and it seems to be fine.
`phpbb_sessions` table size redused like 2,5 MB -> 200 kB, `user_lastvisit` is getting correct.
No issues with login etc.

**Comment by marc1706** [comment]
I've also looked at the query in MySQL and so far it seems alright with regards to using indexes etc.

**Pull request merged**